File: backend/app/routers/characters.py
from fastapi import APIRouter, Depends, File, HTTPException, UploadFile
from sqlalchemy.orm import Session

from app.character_import import parse_character_import
from app.database import get_db
from app.models import Character
from app.schemas import (
    CharacterCreate,
    CharacterOut,
    CharacterUpdate,
    ImportCharacterResponse,
)

router = APIRouter(prefix="/characters", tags=["characters"])


@router.get("", response_model=list[CharacterOut])
def list_characters(db: Session = Depends(get_db)):
    return db.query(Character).order_by(Character.updated_at.desc()).all()


@router.post("", response_model=CharacterOut)
def create_character(payload: CharacterCreate, db: Session = Depends(get_db)):
    character = Character(**payload.model_dump())
    db.add(character)
    db.commit()
    db.refresh(character)
    return character


@router.get("/{character_id}", response_model=CharacterOut)
def get_character(character_id: int, db: Session = Depends(get_db)):
    character = db.query(Character).filter(Character.id == character_id).first()
    if not character:
        raise HTTPException(status_code=404, detail="Character not found")
    return character


@router.patch("/{character_id}", response_model=CharacterOut)
def update_character(
    character_id: int, payload: CharacterUpdate, db: Session = Depends(get_db)
):
    character = db.query(Character).filter(Character.id == character_id).first()
    if not character:
        raise HTTPException(status_code=404, detail="Character not found")
    for key, value in payload.model_dump(exclude_unset=True).items():
        setattr(character, key, value)
    db.commit()
    db.refresh(character)
    return character


@router.delete("/{character_id}")
def delete_character(character_id: int, db: Session = Depends(get_db)):
    character = db.query(Character).filter(Character.id == character_id).first()
    if not character:
        raise HTTPException(status_code=404, detail="Character not found")
    db.delete(character)
    db.commit()
    return {"ok": True}


@router.post("/import", response_model=ImportCharacterResponse)
async def import_character(
    file: UploadFile = File(...), db: Session = Depends(get_db)
):
    content = await file.read()
    try:
        data = parse_character_import(
            filename=file.filename or "character.json", content=content
        )
    except (ValueError, Exception) as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    character = Character(**data)
    db.add(character)
    db.commit()
    db.refresh(character)
    source = "png" if (file.filename or "").lower().endswith(".png") else "json"
    return ImportCharacterResponse(character=character, source=source)

File: crates/server/src/routes/characters.rs
use axum::{
    extract::{Multipart, Path, State},
    routing::{get, post},
    Json, Router,
};
use dreamwell_types::{
    Character, CharacterCreate, CharacterUpdate, ImportCharacterResponse, OkResponse,
};

use crate::character_import::parse_character_import;
use crate::db;
use crate::error::{AppError, AppResult};
use crate::routes::AppState;

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/", get(list_characters).post(create_character))
        .route("/import", post(import_character))
        .route(
            "/:id",
            get(get_character)
                .patch(update_character)
                .delete(delete_character),
        )
}

async fn list_characters(State(state): State<AppState>) -> AppResult<Json<Vec<Character>>> {
    Ok(Json(db::list_characters(&state.pool).await?))
}

async fn create_character(
    State(state): State<AppState>,
    Json(payload): Json<CharacterCreate>,
) -> AppResult<Json<Character>> {
    Ok(Json(db::create_character(&state.pool, payload).await?))
}

async fn get_character(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> AppResult<Json<Character>> {
    Ok(Json(db::get_character(&state.pool, id).await?))
}

async fn update_character(
    State(state): State<AppState>,
    Path(id): Path<i64>,
    Json(payload): Json<CharacterUpdate>,
) -> AppResult<Json<Character>> {
    Ok(Json(db::update_character(&state.pool, id, payload).await?))
}

async fn delete_character(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> AppResult<Json<OkResponse>> {
    db::delete_character(&state.pool, id).await?;
    Ok(Json(OkResponse { ok: true }))
}

async fn import_character(
    State(state): State<AppState>,
    mut multipart: Multipart,
) -> AppResult<Json<ImportCharacterResponse>> {
    let mut filename = "character.json".to_string();
    let mut content = Vec::new();
    while let Some(field) = multipart
        .next_field()
        .await
        .map_err(|e| AppError::bad_request(e.to_string()))?
    {
        if field.name() == Some("file") {
            if let Some(name) = field.file_name() {
                filename = name.to_string();
            }
            content = field
                .bytes()
                .await
                .map_err(|e| AppError::bad_request(e.to_string()))?
                .to_vec();
        }
    }
    let payload = parse_character_import(&filename, &content)?;
    let character = db::create_character(&state.pool, payload).await?;
    let source = if filename.to_lowercase().ends_with(".png") {
        "png"
    } else {
        "json"
    };
    Ok(Json(ImportCharacterResponse {
        character,
        source: source.to_string(),
    }))
}
